yarn-project/simulator/src/public/public_tx_simulator.ts

```diff
@@ -0,0 +1,263 @@
+        context.state.discardForkedState();
+      } else {
+        if (!context.hasPhase(TxExecutionPhase.TEARDOWN)) {
+          // Nothing to do after this (no teardown), so merge state in now instead of letting teardown handle it.
+          context.state.mergeForkedState();
+        }
+      }
+
+      return result;
+    }
+  }
+
+  private async processTeardownPhase(context: PublicTxContext): Promise<ProcessedPhase | undefined> {
+    context.progressToNextPhase(); // to teardown
+    if (context.hasPhase()) {
+      if (!context.state.isForked()) {
+        // if state isn't forked (app logic was empty or reverted), fork now
+        // so we can rollback to the end of setup on teardown revert
+        context.state.fork();
+      }
+
+      const result = await this.processPhase(context);
+
+      if (result.reverted) {
+        // Drop the currently active forked state manager and rollback to end of setup.
+        context.state.discardForkedState();
+      } else {
+        context.state.mergeForkedState();
+      }
+
+      return result;
+    }
+  }
+
+  private async processPhase(context: PublicTxContext): Promise<ProcessedPhase> {
+    const tx = context.tx;
+    const callRequests = context.getCallRequestsForCurrentPhase();
+    const executionRequests = context.getExecutionRequestsForCurrentPhase();
+    const txStateManager = context.state.getActiveStateManager();
+
+    this.log.debug(
+      `Beginning processing in phase ${TxExecutionPhase[context.getCurrentPhase()]} for tx ${tx.getTxHash()}`,
+    );
+
+    const returnValues: NestedProcessReturnValues[] = [];
+    let reverted = false;
+    let revertReason: SimulationError | undefined;
+    const phaseTimer = new Timer();
+    for (let i = callRequests.length - 1; i >= 0; i--) {
+      if (reverted) {
+        break;
+      }
+
+      const callRequest = callRequests[i];
+      const executionRequest = executionRequests[i];
+
+      // add new contracts to the contracts db so that their functions may be found and called
+      // TODO(#4073): This is catching only private deployments, when we add public ones, we'll
+      // have to capture contracts emitted in that phase as well.
+      // TODO(@spalladino): Should we allow emitting contracts in the fee preparation phase?
+      // TODO(#6464): Should we allow emitting contracts in the private setup phase?
+      // if so, this should only add contracts that were deployed during private app logic.
+      // FIXME: we shouldn't need to directly modify worldStateDb here!
+      await this.worldStateDB.addNewContracts(tx);
+
+      // each enqueued call starts with an incremented side effect counter
+      // FIXME: should be able to stop forking here and just trace the enqueued call (for hinting)
+      // and proceed with the same state manager for the entire phase
+      const enqueuedCallStateManager = txStateManager.fork(/*incrementSideEffectCounter=*/ true);
```

Comment:
This note is important and we can fix this once we stop calling kernel merge. We should not need to fork state for every enqueued call. We should be able to work with the same state manager for an entire phase.